File: 0007-reverse-integer/0007-reverse-integer.cpp
class Solution {
public:
    int reverse(int x) {
        if (x == numeric_limits<int>::max() || x == numeric_limits<int>::min()) {
            return 0;
        }

        int mult = x < 0 ? -1 : 1;
        x = std::abs(x);

        int res = 0;

        while (x > 0) {
            if (res > numeric_limits<int>::max() / 10) {
                return 0;
            }

            res *= 10;

            int dig = x % 10;
            x /= 10;
            
            if (res + dig < res) {
                return 0;
            }

            res += dig;
        }

        if (mult == -1 && res == numeric_limits<int>::max()) {
            return 0;
        }

        return res * mult;
    }
};

File: 0007-reverse-integer/0007-reverse-integer.cs
public class Solution {
    public int Reverse(int x) {
        if (x == Int32.MinValue || x == Int32.MaxValue) {
            return 0;
        }

        int result = 0;
        bool isNeg = x < 0;
        x = Math.Abs(x);

        while (x > 0) {
            int dig = x % 10;
            x /= 10;

            if (result > Int32.MaxValue / 10) {
                return 0;
            }

            result *= 10;

            if (result + dig < result) {
                return 0;
            }

            result += dig;
        }

        if (isNeg) {
            result *= -1;
        }

        return result;
    }
}